### PR TITLE
experiment with cupy and pycuda

### PR DESCRIPTION
**Pull request by afif-ishamsyah**
There are problems with using cupy and pycuda

### CODE REVIEW

**Comment by SimeonEhrig** [comment]
`cupy_pybind2.py` has the right direction. But instead using a global C variable for the device memory, we need a local variable which we can handle in the python code and pass through the pybind interfaces. Did you tried to use cupy arrays like numpy arrays together with pybind11? If it works, there is a nice way for the memory handling.

Maybe you should also skip multi GPU for the moment. First develop a solution for a single GPU and then extend to multi GPU.

I found a nice example in the cupy documentation: https://docs.cupy.dev/en/stable/user_guide/kernel.html#raw-kernels
You can take the example and replace the `add_kernel` variable with your pybind11 binding. If this works, we can extend it with everything, what we need.

**Comment by afif-ishamsyah** [comment]
So we use RawKernel to pybind instead of __global__ kernel, and then using it to allocate memory?

**Comment by SimeonEhrig** [comment]
No, I mean, take the following code:

```python
>>> import cupy as cp
add_kernel = cp.RawKernel(r'''
... extern "C" __global__
... void my_add(const float* x1, const float* x2, float* y) {
...     int tid = blockDim.x * blockIdx.x + threadIdx.x;
...     y[tid] = x1[tid] + x2[tid];
... }
... ''', 'my_add')
>>> x1 = cp.arange(25, dtype=cp.float32).reshape(5, 5) # gpu memory allocation with python
>>> x2 = cp.arange(25, dtype=cp.float32).reshape(5, 5)
>>> y = cp.zeros((5, 5), dtype=cp.float32)
>>> add_kernel((5,), (5,), (x1, x2, y))  # grid, block and arguments
>>> y
array([[ 0.,  2.,  4.,  6.,  8.],
       [10., 12., 14., 16., 18.],
       [20., 22., 24., 26., 28.],
       [30., 32., 34., 36., 38.],
       [40., 42., 44., 46., 48.]], dtype=float32)
```

and transform it to:

```python
>>> import cupy as cp
>>> import myPythonBind
>>> x1 = cp.arange(25, dtype=cp.float32).reshape(5, 5) # gpu memory allocation with python
>>> x2 = cp.arange(25, dtype=cp.float32).reshape(5, 5)
>>> y = cp.zeros((5, 5), dtype=cp.float32)
>>>  myPythonBind(x1, x2, y) # kernel and kernel launch are written in C++ and has a pybind11 binding
>>> y
array([[ 0.,  2.,  4.,  6.,  8.],
       [10., 12., 14., 16., 18.],
       [20., 22., 24., 26., 28.],
       [30., 32., 34., 36., 38.],
       [40., 42., 44., 46., 48.]], dtype=float32)
```

**Comment by afif-ishamsyah** [comment]
so the "kernel and kernel launch are written in C++ and has a pybind11 binding" part is for the partial_update part?

**Comment by SimeonEhrig** [comment]
Yes. But in the beginning, you should focus if it possible to pass the cupy array through the pybind11 interface. If I understand it correctly, the cupy array is the central memory object of cupy.

**Comment by afif-ishamsyah** [comment]
that is the hardest part because most of the time all I get is segmetation fault

**Comment by SimeonEhrig** [comment]
But this is the most interesting part. I think, I found the reason of the segmentation faults: https://github.com/pybind/pybind11/issues/2694

And here is a workaround: https://stackoverflow.com/questions/66989716/passing-cupy-cuda-device-pointer-to-pybind11

**Comment by afif-ishamsyah** [comment]
Oh, I mean I always get segmentation fault when returning the array back to cupy. For example in commit b5a6c778b6ba000e8c208edbdb81cab1446ee8d5 (2 commit before this), in gpu_algo.hpp at line 64. 

For the workaround link, you can see the gpu_algo.hpp that I already done something similiar everytime I receive an array from python.

**Comment by SimeonEhrig** [comment]
> Oh, I mean I always get segmentation fault when returning the array back to cupy. For example in commit [b5a6c77](https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/commit/b5a6c778b6ba000e8c208edbdb81cab1446ee8d5) (2 commit before this), in gpu_algo.hpp at line 64.
> 
In this [line](https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/b5a6c778b6ba000e8c208edbdb81cab1446ee8d5/gpu_memory_management/cupy_pybind.py#L68) of code, I see two problems. The first is, that you don't send back an array. It's just a pointer. Can you please run `type(gpu_image)` to check the Python type. I'm not sure, how pointers are represented in Python. Second, you executed a print on GPU memory. This also causes an segmentation fault in C++ application, because you tried to access GPU memory directly from the CPU.

**Comment by afif-ishamsyah** [comment]
I check it again and segmantation fault always happens everytime I use return to a CUDA variable

**Comment by SimeonEhrig** [comment]
I think the main problems is, that Python does not support pointers. In this post, it is mentioned, that a raw pointer is casted to a single value: https://stackoverflow.com/questions/57990269/passing-pointer-to-c-from-python-using-pybind11

I think we, need a wrapper object, like we already have with the numpy array for the CPU side. In the post, the class `py::buffer` was suggested. Writing a own wrapper class is also possible or using C++ [smart pointers](https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html), but I would no suggest this, because smart pointers has the same problem like raw pointers, we have enough information about the data, like the length.

**Comment by SimeonEhrig** [comment]
The last commit is the right direction. [Allocate GPU memory](https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/d99003f3bbf8ba911e58e517c283ee3afcde77e5/gpu_memory_management/cupy_pybind4.py#L30) on the Python side and [use it](https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/d99003f3bbf8ba911e58e517c283ee3afcde77e5/gpu_memory_management/cupy_pybind.cu#L50) on the C++ side.

Only this cast looks ugly: https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/d99003f3bbf8ba911e58e517c283ee3afcde77e5/gpu_memory_management/cupy_pybind.cu#L50
and the Python interface is not so nice but it works. I think we will find a better solution in future. At the moment I check, if we can implement support for cupy arrays like numpy arrays. In theory, it should be possible.

Something about memory management in cupy. The first time it can be confusing for a C++ developer. The lifetime of memory is bound to the cupy arrays. If you delete the reference to the object, the memory is unbound, e.g.:

```python
import cupy

z = cupy.zeros(1024*1024*1024)
# delete array and unbound the GPU memory 
del z
k = cupy.zeros(1024*1024*1024)
# the array is implicit delete and the GPU memory unbound, because k does not reference to the array anymore 
k = 1
```
But unbound does not means delete. The `nvidia-smi` will still show used memory. The reason is the memory manager of cupy. Instead calling the `cudaFree()` function, the manager marked the memory as unused and reuse it later again. This is more efficient than calling `cudaFree()` and `cudaMalloc` again. This line of code only forces the memory manager to free all unused memory:

https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/d99003f3bbf8ba911e58e517c283ee3afcde77e5/gpu_memory_management/cupy_pybind4.py#L11

But in practice, it is not necessary.

I think, your next step should making your example multi-GPU capable. Means allocate memory and execute a kernel on a specific GPU by it's id. In the meantime, I check the requirements for the native cupy array support.

File: gpu_memory_management/cupy_pybind.py [hunk 1]
```diff
@@ -39,31 +20,41 @@ def free_gpu_memory(number_of_gpus):
...
-  print(partial_update)
+  cp.cuda.Device(0).use()
+  gpu_image_0 = cp.zeros(parted_image[0].size)
+  gpu_partial_update_0 = cp.zeros(parted_image[0].size)
 
-  # allocate memory for each GPU, the size is parted image size + update value size + parted update value size (?)
-  # need to know how to properly set the size
-  print("Number of Device : ", number_of_gpus)
-  # alloc_gpu_memory(parted_images, number_of_gpus)
+  cp.cuda.Device(1).use()
```
Comment (SimeonEhrig) [inline gpu_memory_management/cupy_pybind.py:33]: Can you make the example independent of the number of GPU's please. Means, ask for the number of available GPU's in the beginning and iterate over the number.

Comment (afif-ishamsyah) [reply gpu_memory_management/cupy_pybind.py:33]: So number of GPU is a user input?

Comment (SimeonEhrig) [reply gpu_memory_management/cupy_pybind.py:33]: No, this is not necessary. Check it in the source code, how many GPU's are available. You can restrict the number of visible CUDA GPU's via environment variable: `CUDA_VISIBLE_DEVICES=0 python example.py`.

Comment (afif-ishamsyah) [reply gpu_memory_management/cupy_pybind.py:33]: I already check the number of GPU in line 17 in cupy_pybind.py

Comment (SimeonEhrig) [reply gpu_memory_management/cupy_pybind.py:33]: Okay, but you don't use the variable all the time. If there is only a single GPU available, this line throws an error:

https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/99fb7e60f171c766dfaff9c1b6fd743c97ad833d/gpu_memory_management/cupy_pybind.py#L33

File: gpu_memory_management/test/test.hpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#include <string>
+#include <random>
+
+#define CUDA_CHECK(call) {cudaError_t error = call; if(error!=cudaSuccess){printf("<%s>:%i ",__FILE__,__LINE__); printf("[CUDA] Error: %s\n", cudaGetErrorString(error));}}
+using namespace std;
+using namespace std::literals::complex_literals;
+namespace py = pybind11;
+
+//test if a pointer is a device/gpu pointer or not
+int is_device_pointer(const void *ptr)
```
Comment (SimeonEhrig) [inline gpu_memory_management/test/test.hpp:25]: For some of your tests, you can use the return value `bool`, because the test result can be only correct or not. This allows also to write more clean code, e.g.:

```c++
bool is_device_pointer(const void *ptr){
  cudaPointerAttributes attributes;

  CUDA_CHECK(cudaPointerGetAttributes(&amp;attributes, ptr));

  //according to documentation, if cudaMemoryTypeDevice = 2, then it is a device memory
  // will return true, if device memory
  return attributes.type == 2;
}
```

File: gpu_memory_management/test/test.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import pytest
```
Comment (SimeonEhrig) [inline gpu_memory_management/test/test.py:1]: Just as hint. Pytest executes all files, stating with the prefix `test_` automatically in the folder. So if you rename your `test.py` to `test_<something>.py` you can execute `pytest` in the build folder and pytest will execute all tests in the file automatically.

File: gpu_memory_management/test/test.py [hunk 4]
```diff
@@ -7,68 +7,90 @@
...
     assert(res == 1)
 
-#test 6 : copy array of ones from real cupy to cpu
-def test_copy_cupy_of_ones_to_cpu():
-    a = cp.ones(5)
-    b = gpuMemManagement.test_copy_cupy_of_ones_to_cpu(a.data.ptr, a.size)
+#test 6 : copy array of float from real cupy to cpu
+def test_copy_real_cupy_to_cpu():
+    cp.cuda.Device(0).use()
+    a = cp.array([3.14, 3.14, 3.14])
```
Comment (SimeonEhrig) [inline gpu_memory_management/test/test.py:52]: I strongly recommend to not using a simple pattern for your test data, like repeating the same number or count from 0 until n. That data are corner cases and can provide different problems. For example, you want to implement a loop, which iterate over all elements of an array and do something with it. If you have the same value in all cells of your array, it is possible that you do a mistake in the loop and instead you loop over each element, you only loops over the first element N times (this happens more then one time for me ;-)). For the corner case of repeating numbers, the result is the same, but if you have different elements in the array, your algorithm will fail.

But this does not mean, that you need to replace all test data in all tests. If you only need a valid pointer or wants to check the size of an array, `np.ones()` is fine.

File: gpu_memory_management/test/test.py [hunk 5]
```diff
@@ -7,68 +7,90 @@
...
+    assert(c == 3*(4.5 + 1))
 
-#test 9 : test 2 reinterpret cast with different data type and see if both has a same value
-#this part need more explanation
-def test_2_different_reiterpret_cast():
-    a = cp.array([3.14, 3.14, 3.14])
-    b = gpuMemManagement.test_2_different_reiterpret_cast(a.data.ptr)
+#one of the reason why test 1 - 9 was successfull is because all cupy array value in the test is always double precision python float,
+#the reinterpret cast of the real cupy is always casted to double, and the custom cupy current design is only accepting double type.
+#test 10 : what if instead of python float, array of integer is sent by real cupy. expectation : failure
```
Comment (SimeonEhrig) [inline gpu_memory_management/test/test.py:83]: That's make totally sense. By default numpy uses `numpy.float64`, if you create a new array with python floating pointer numbers, which is equal to C++ `double`. If you want to create a numpy array with `numpy.float32` (which is equal to C++ `float`), you have to set the element type manual: `a = np.array([3.0, 4.5], dtype=np.float32)` 

Example: https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/39c854178a1f7717667f17c41f6946347bb3496f/cupy_array_caster/test/0004_pointer_caster/test_pointerCaster.py#L77

File: gpu_memory_management/test/test_custom_cupy.py [hunk 6]
```diff
@@ -77,15 +77,28 @@ def test_custom_cupy_pointer_with_cuda_kernel():
...
     a = cp.array([3.14, 4.25, 5.36])
     b = cupy_ref.Custom_Cupy_Ref(ptr = a.data.ptr, size = a.size)
     c = gpuMemManagement.test_copy_custom_cupy_to_custom_cupy(b)
 
     assert(b.size == c.size and b.ptr == c.ptr)
+
+#test 12 : test memory usage, still not sure if this is a right way
+def test_memory():
+    assert(cp.get_default_memory_pool().used_bytes() == 0)
+    test_create_real_cupy_from_c()
```
Comment (SimeonEhrig) [inline gpu_memory_management/test/test_custom_cupy.py:102]: You should at least assign the return value to a Python variable to give Python securely the ownership over the memory.

I suggest the following extension of the test:
```Python
assert(cp.get_default_memory_pool().used_bytes() == 0)

a = test_create_real_cupy_from_c()
# check, if returned memory works like expected
b = a + cp.array([1.0, 2.0, 3.0])

# This is a simplified version of the assert. I think you need a little bit more code for the compare.
assert b == expected_result

# delete cupy arrays
a = None
b = None

assert(cp.get_default_memory_pool().used_bytes() == 0)
```

Comment (afif-ishamsyah) [reply gpu_memory_management/test/test_custom_cupy.py:102]: Okay. I just realized that I forgot to give a return value to the python. I will fix this test

File: gpu_memory_management/cupy_pybind.cu [hunk 7]
```diff
@@ -39,6 +39,26 @@ PYBIND11_MODULE(gpuMemManagement, m)
...
+  //missmatch error always appeared based on the order of types
+  //ex : if cupy with double type is used, them there will be 3 missmatch error for uint16, uint32, and float
+  //ex : if cupy with complex128 type is used, them there will be 5 missmatch error for uint16, uint32, float, double, and complex64
+  //this error appeard with "pytest -s"
+  m.def("test_custom_cupy_template_function", &test_custom_cupy_template_function<uint16_t>);
+  m.def("test_custom_cupy_template_function", &test_custom_cupy_template_function<uint32_t>);
+  m.def("test_custom_cupy_template_function", &test_custom_cupy_template_function<float>);
+  m.def("test_custom_cupy_template_function", &test_custom_cupy_template_function<double>);
+  m.def("test_custom_cupy_template_function", &test_custom_cupy_template_function<complex<float>>);
+  m.def("test_custom_cupy_template_function", &test_custom_cupy_template_function<complex<double>>);
```
Comment (SimeonEhrig) [inline gpu_memory_management/cupy_pybind.cu:52]: Do you know what happens, if you overload a python function. Normally, it is forbidden. I don't know, what pybind11 is doing.

Comment (afif-ishamsyah) [reply gpu_memory_management/cupy_pybind.cu:52]: I do it in the previous phase retrieval project and it is working fine

Comment (SimeonEhrig) [reply gpu_memory_management/cupy_pybind.cu:52]: But in the last project, you didn't used custom caster and I'm not sure, if you also didn't used templates. You should test it without overloading. It is a cheep try.

Comment (afif-ishamsyah) [reply gpu_memory_management/cupy_pybind.cu:52]: I already tried without overloading before this commit. Without overloading, I get a pybind11 error. I look into it and I found that I need an overloading
